### PR TITLE
Fix urlretrieve issue on python2

### PR DESCRIPTION
**Pull request by hackalog**
Occasionally, when calling `urlretrieve()` on a url, a strange error is seen  (#322) about a missing `readinto`. I think this is a bug in python-future (https://github.com/PythonCharmers/python-future/pull/423) on Python 2. In the meantime, my workaround is to require and use the `requests` module in kivy-ios in this failure case.

While here, incorporate @JonasT's retry idea proposed in https://github.com/kivy/python-for-android/pull/1574

### CODE REVIEW

**Comment by JonasT** [comment]
I'm not involved with kivy-ios, but it would probably be helpful to explain why you are moving certain files to another location. Also, catching *all* exceptions from `urlretrieve` seems like overkill *at least to me* - it might be preferable to investigate why you're actually getting this `AttributeError` because it seems like a requests or stdlib bug that should probably be fixed at the right location

**Comment by hackalog** [comment]
The file moves are because this diff is build on top of #331 (required to even get to the point where this bug is triggered), and so those parts will go away if and when #331 is is merged (and can be ignored for now).

The bug being addressed is the code path that leads to the upstream bug in python-future, which is getting invoked here. I see no easy fix unless upstream accepts https://github.com/PythonCharmers/python-future/pull/423

That being said, I should mark this a work in progress, at least for now

**Comment by JonasT** [comment]
@hackalog ah, but if it's such a specific known issue, then you should definitely just catch `AttributeError`, and only if Python 2 is used! (check with `sys.version_info.major`) Also I highly recommend adding a code comment to that `except` clause which links the bug you referenced. Otherwise, everyone will forget why on earth this code is catching an AttributeError and when that could ever be possibly removed (assuming upstream ever fixes it)

**Comment by hackalog** [comment]
Those suggestions are all legit. I had originally thought a drop-in one line replacement of requests for urlretrieve would do the trick, but FTP reared its ugly head.

Let me work this up into a less heavy-handed form.

**Comment by JonasT** [comment]
that looks like a nice combination :smile: I like how it looks like now

**Comment by hackalog** [comment]
Yes, thanks for the feedback @JonasT . As far as crude workarounds go, it's looking almost forgivable now. :)

File: toolchain.py [hunk 1]
```diff
@@ -18,16 +18,18 @@
...
     from urllib.request import FancyURLopener, urlcleanup
 except ImportError:
     from urllib import FancyURLopener, urlcleanup
+
 try:
     from pbxproj import XcodeProject
     from pbxproj.pbxextensions.ProjectFiles import FileOptions
 except ImportError:
-    print("ERROR: pbxproj requirements is missing")
+    print("ERROR: Python requirements are missing")
```
[the remark this answers is not in the document]
Comment (hackalog) [reply toolchain.py]: Ah, an interim version of the fix added the requests import here, so there were 2 packages being imported (https://github.com/learnleapfly/kivy-ios/commit/999dc68123a148b92af914b2f9acaed4789572f2#diff-c3327fdae2ebcc6a34dc5a9d23ae9ffd) I neglected to change it back (sort of figuring more things might end up in requirements.txt before long)

**Comment by hackalog** [comment]
@tito would you like me to change that Python->pbxproj string back, or does this look good to merge?

**Pull request merged**